eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggerer.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2022 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment.relay
+
+import akka.actor.typed.ActorRef.ActorRefOps
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.typed.{ActorRef, Behavior, SupervisorStrategy}
+import fr.acinq.bitcoin.scalacompat.ByteVector32
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.BlockHeight
+import fr.acinq.eclair.blockchain.CurrentBlockHeight
+import fr.acinq.eclair.io.PeerReadyNotifier.NotifyWhenPeerReady
+import fr.acinq.eclair.io.{PeerReadyNotifier, Switchboard}
+import fr.acinq.eclair.payment.relay.AsyncPaymentTriggerer.Command
+
+import scala.concurrent.duration.Duration
+
+/**
+ * This actor waits for an async payment receiver to become ready to receive a payment or for a block timeout to expire.
+ * If the receiver of the payment is a connected peer, spawn a PeerReadyNotifier actor.
+ * TODO: If the receiver is not a connected peer, wait for a `ReceiverReady` onion message containing the specified paymentHash.
+ */
+
+object AsyncPaymentTriggerer {
+  // @formatter:off
+  sealed trait Command
+  case class Start(switchboard: ActorRef[Switchboard.GetPeerInfo]) extends Command
+  case class Watch(replyTo: ActorRef[Result], remoteNodeId: PublicKey, paymentHash: ByteVector32, timeout: BlockHeight) extends Command
+  private case class WrappedPeerReadyResult(result: PeerReadyNotifier.Result) extends Command
+  private case class WrappedCurrentBlockHeight(currentBlockHeight: CurrentBlockHeight) extends Command
+
+  sealed trait Result
+  case object AsyncPaymentTriggered extends Result
+  case object AsyncPaymentTimeout extends Result
+  // @formatter:on
+
+  def apply(): Behavior[Command] = Behaviors.setup { context =>
+    new AsyncPaymentTriggerer(context).initializing()
+  }
+}
+
+private class AsyncPaymentTriggerer(context: ActorContext[Command]) {
+  import AsyncPaymentTriggerer._
+
+  case class Watcher(replyTo: ActorRef[Result], timeout: BlockHeight, paymentHash: ByteVector32) {
+    def expired(currentBlockHeight: BlockHeight): Boolean = timeout <= currentBlockHeight
+  }
+  case class AsyncPaymentTrigger(notifier: ActorRef[PeerReadyNotifier.Command], watchers: Set[Watcher]) {
```

Comment:
I struggled at first because the names of these two classes are quite misleading. The `Watcher` is actually just a payment, and the `AsyncPaymentTrigger` is just a remote peer for which we have pending payments. I renamed these two classes in https://github.com/ACINQ/eclair/commit/ac5118b0b5dc1c1f543bca076a43aba5f4ea9ef3, let me know what you think.

Reply:
Yes - I agree this is a clearer naming scheme.